packages/config/src/class-config.php

```diff
@@ -51,52 +51,113 @@ public function ensure( $feature ) {
+				&& $this->ensure_feature( 'tracking' );
+		}
 
-			// Check for WP Super Cache.
-			$this->config['sync_wp_super_cache'] ? Sync_Actions::initialize_wp_super_cache() : null;
+		if ( $this->config['sync'] ) {
+			$this->ensure_class( 'Automattic\Jetpack\Sync\Main' )
+				&& $this->ensure_feature( 'sync' );
 		}
 	}
 
 	/**
-	 * Runs on default plugins_loaded hook priority.
+	 * Returns true if the required class is available and alerts the user if it's not available
+	 * in case the site is in debug mode.
 	 *
-	 * @action plugins_loaded
+	 * @param String $classname a fully qualified class name.
+	 * @return Boolean whether the class is available.
 	 */
-	public function on_plugins_loaded() {
-		if ( $this->config['tos'] && $this->config['tracking'] ) {
-			$terms_of_service = new Terms_Of_Service();
-			$tracking         = new Plugin_Tracking();
-			if ( $terms_of_service->has_agreed() ) {
-				add_action( 'init', array( $tracking, 'init' ) );
-			} else {
-				/**
-				 * Initialize tracking right after the user agrees to the terms of service.
-				 */
-				add_action( 'jetpack_agreed_to_terms_of_service', array( $tracking, 'init' ) );
-			}
+	protected function ensure_class( $classname ) {
+		$available = class_exists( $classname );
+
+		if ( ! $available && defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+			trigger_error( // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_trigger_error
+				sprintf(
+					/* translators: %1$s is a PHP class name, %2$s is a Jetpack feature code, like 'sync'. */
+					esc_html__(
+						'Unable to load class %1$s, not enabling %2$s feature. Please add the package that contains it using composer and make sure you are requiring the Jetpack autoloader',
+						'jetpack'
+					),
+					esc_html( $classname )
+				),
+				E_USER_NOTICE
+			);
 		}
+
+		return $available;
 	}
 
 	/**
-	 * Runs after most of plugins_loaded hook functions have been run.
+	 * Ensures a feature is enabled, sets it up if it hasn't already been set up.
 	 *
-	 * @action plugins_loaded
+	 * @param String $feature slug of the feature.
+	 * @return Boolean whether the feature was enabled. If it was already enabled, returns false.
 	 */
-	public function on_plugins_loaded_late() {
-		/*
-		 * Init after plugins loaded and before the `init` action. This helps with issues where plugins init
-		 * with a high priority or sites that use alternate cron.
+	protected function ensure_feature( $feature ) {
+		if ( did_action( 'jetpack_feature_' . $feature . '_enabled' ) ) {
+			return false;
+		}
+
+		$method = 'enable_' . $feature;
+		if ( ! method_exists( $this, $method ) ) {
+			return false;
+		}
+
+		$this->{ $method }();
+
+		/**
+		 * Fires when a specific Jetpack package feature is initalized using the Config package.
+		 *
+		 * @since 8.1.0
 		 */
-		Sync_Actions::init();
+		do_action( 'jetpack_feature_' . $feature . '_enabled' );
 	}
+
+	/**
+	 * Dummy method to enable Terms of Service.
+	 */
+	protected function enable_tos() {
```

Comment:
Should we hardcode in these features or do something more dynamic?

Each package that needed to be enabled adds some basic `enable` function that we then fire, then we just call `( new {class_name})->enable()` or something to that effect?

Thinking that a package itself, if there is a way to "enable" it is better placed to know what needs to be done vs having it hardcoded here?

Reply:
I agree here that something more dynamic would be really nice, but it's a next step in my opinion. I'd rather start moving things to the configuration class for now, and later see how it all works with different init scenarios. Already we can see that features have dependencies, and we'll need to see how they play together when moved to that passive init mode.

Reply:
That's fair.